faiss/gpu/impl/RaftIVFPQ.cu

```diff
@@ -0,0 +1,564 @@
+        bool storePairs) {
+    // TODO: Fill this in!
+}
+
+size_t RaftIVFPQ::getGpuListEncodingSize_(idx_t listId) {
+    return static_cast<size_t>(
+            raft_knn_index.value().get_list_size_in_bytes(listId));
+}
+
+/// Return the encoded vectors of a particular list back to the CPU
+std::vector<uint8_t> RaftIVFPQ::getListVectorData(idx_t listId, bool gpuFormat)
+        const {
+    if (gpuFormat) {
+        FAISS_THROW_MSG(
+                "gpuFormat should be false for RAFT indices. Unpacked codes are flat.");
+    }
+    FAISS_ASSERT(raft_knn_index.has_value());
+
+    const raft::device_resources& raft_handle =
+            resources_->getRaftHandleCurrentDevice();
+    auto stream = raft_handle.get_stream();
+
+    idx_t listSize = getListLength(listId);
+
+    auto cpuListSizeInBytes = getCpuVectorsEncodingSize_(listSize);
+
+    std::vector<uint8_t> flat_codes(
+            cpuListSizeInBytes, static_cast<uint8_t>(0));
+
+    idx_t maxBatchSize = 65536;
+    for (idx_t offset_b = 0; offset_b < listSize; offset_b += maxBatchSize) {
+        uint32_t batchSize = min(maxBatchSize, listSize - offset_b);
+        uint32_t bufferSize = getCpuVectorsEncodingSize_(batchSize);
+        uint32_t codesOffset = getCpuVectorsEncodingSize_(offset_b);
+
+        // Fetch flat PQ codes for the current batch
+        auto codes_d = raft::make_device_vector<uint8_t>(
+                raft_handle, static_cast<uint32_t>(bufferSize));
+
+        raft::neighbors::ivf_pq::helpers::unpack_contiguous_list_data(
+                raft_handle,
+                raft_knn_index.value(),
+                codes_d.data_handle(),
+                batchSize,
+                listId,
+                offset_b);
+
+        // Copy the flat PQ codes to host
+        raft::update_host(
+                flat_codes.data() + codesOffset,
+                codes_d.data_handle(),
+                bufferSize,
+                stream);
+        raft_handle.sync_stream();
+    }
+
+    return flat_codes;
+}
+
+/// Find the approximate k nearest neighbors for `queries` against
+/// our database
+void RaftIVFPQ::search(
+        Index* coarseQuantizer,
+        Tensor<float, 2, true>& queries,
+        int nprobe,
+        int k,
+        Tensor<float, 2, true>& outDistances,
+        Tensor<idx_t, 2, true>& outIndices) {
+    uint32_t numQueries = queries.getSize(0);
+    uint32_t cols = queries.getSize(1);
+    idx_t k_ = std::min(static_cast<idx_t>(k), raft_knn_index.value().size());
+
+    // Device is already set in GpuIndex::search
+    FAISS_ASSERT(raft_knn_index.has_value());
+    FAISS_ASSERT(numQueries > 0);
+    FAISS_ASSERT(cols == dim_);
+    FAISS_THROW_IF_NOT(nprobe > 0 && nprobe <= numLists_);
+
+    const raft::device_resources& raft_handle =
+            resources_->getRaftHandleCurrentDevice();
+    raft::neighbors::ivf_pq::search_params pams;
+    pams.n_probes = nprobe;
+    pams.lut_dtype = useFloat16LookupTables_ ? CUDA_R_16F : CUDA_R_32F;
```

Comment:
Do we plan to have a way to enable `lut_type = CUDA_R_8U`?

Reply:
We must adhere to the FAISS public API. The GpuIndexIVFPQ index initializations do not support 8 bit lookup tables, so not adding it for the RAFT side also.

Reply:
@tarang-jain this is a pretty important feature that RAFT has added to the IVF-PQ implementation. We should definitely include this in a follow-up.

Reply:
Created tracking issue #3249